ostd/src/task/processor.rs

```diff
@@ -96,16 +90,60 @@ pub(super) fn switch_to_task(next_task: Arc<Task>) {
         context_switch(current_task_ctx_ptr, next_task_ctx_ptr);
     }
 
-    // Now it's fine to drop `prev_task`. However, we choose not to do this because it is not
-    // always possible. For example, `context_switch` can switch directly to the entry point of the
-    // next task. Not dropping is just fine because the only consequence is that we delay the drop
-    // to the next task switching.
-
-    // See also `kernel_task_entry`.
-    crate::arch::irq::enable_local();
+    on_task_entry_or_resume();
 
-    // The `next_task` was moved into `CURRENT_TASK_PTR` above, now restore its FPU state.
     if let Some(current) = Task::current() {
         current.restore_fpu_state();
     }
 }
+
+fn update_prev_task_ptr(current_task_ptr: *const Task) {
+    #[cfg(debug_assertions)]
+    {
+        let old_prev = PREVIOUS_TASK_PTR.load();
+        debug_assert!(old_prev.is_null());
+    }
+    PREVIOUS_TASK_PTR.store(current_task_ptr);
+}
+
+fn before_switching_to(next_task: &Task, irq_guard: &DisabledLocalIrqGuard) {
+    // Ensure that the mapping to the kernel stack is valid.
+    next_task.kstack.flush_tlb(irq_guard);
+
+    // Ensure that we are not switching to a task that is already running.
+    while next_task
+        .switched_to_cpu
+        .compare_exchange(false, true, Ordering::AcqRel, Ordering::Relaxed)
+        .is_err()
+    {
+        log::warn!("Switching to a task already running in the foreground");
+        core::hint::spin_loop();
+    }
+}
+
+pub(super) fn on_task_entry_or_resume() {
```

Comment:
This should be an `unsafe fn` because this method can only be called once after each context switch. Otherwise there will be soundness problems. (For example, it may enable the IRQ while some IRQ guards are present).